### ISSUE TITLE
Implement ScopeProvider and LinkingService

### ISSUE THREAD

**Issue by szarnyasg** [slizaa/slizaa-opencypher-xtext#7]
With @jmarton, we are still using this grammar for our [ingraph](https://github.com/FTSRG/ingraph) research project. Lately, we've been working on our query compiler to support `WITH` clauses and aliases. We realized that a lot of the scoping/linking/name resolution could (and probably should) be handled by Xtext, so I started to work on this. However, it turned out that this aspect of the (open)Cypher language is quite tricky as this simple query demonstrates:

![image](https://user-images.githubusercontent.com/1402801/34397633-fa789fdc-eb77-11e7-8827-9e0d02be7daf.png)

(The dashed line represents that the purple `p` alias will refer to the green `p` , but these are two different variables.)

Some lessons I learnt so far:

1. In total, there are 4 clauses that can introduce new variables: `MATCH`, `UNWIND`, `WITH` and `RETURN`.
1. There are multiple ways to approach scopes. We could create a scope within the (transitively) containing `SingleQuery` of the variables. Or we could created nested scopes that keep growing from top to the bottom. In this case, scopes are non-trivial as illustrated by this query, where `ORDER BY` uses both variable `x` and its alias, `y` in `RETURN`, along with `z` which is not returned and `w` which is just introduced by the `RETURN` clause.

   ```
   WITH {p1: 'some', p2: 'thing'} AS x, 'dummy' AS z
   RETURN x AS y, 'smart' AS w
   ORDER BY x.p1, y.p2, z, w
   ```
1. One could argue that Cypher mixes declarations and usages in the `MATCH` clauses. For example, in `MATCH (n), (n) RETURN n`, the first variable `n` is a declaration, but the second `n` is a usage of this variable and so is the third. So ideally, when refactoring the second `n` value, all of them should change. However, this is not critical for our use case, as we can handle this in our query the compiler (and the compiler does not care about refactoring anyways).
1. I also checked the latest [Graph Database plug-in](https://plugins.jetbrains.com/plugin/8087-graph-database-support) for IntelliJ, but it did not seem to handle these issues particularly well, e.g. it did not refactor the example query in the figure correctly.

Here is my current progress: https://github.com/szarnyasg/slizaa-opencypher-xtext/pull/1/files. I did my fair share of Eclipse development, but I only have limited experience with Xtext, so any feedback is welcome. If I can get this to a working state, I'd be happy to contribute it to Slizaa. @wuetherich, do you think such a feature would be a valuable addition to the Slizaa openCypher grammar?

**Comment by wuetherich** [comment]
Hi Gabor,

Happy new year and sorry for the delayed reply, I've been on vacation for some days with my family... 

To answer you question: Yes, scoping support definitely would be a useful and valueable addition to slizaa-opencypher-xtext. I also stumble about the scoping issues when 
I implemented the grammar but haven't haven't found yet the time to investigate it further. So any help is more than welcome!

Btw, currently the opencypher-xtext implementation is part of my slizaa project, but maybe it would make sense to promote it bit more and provide it in its own, stand-alone toplevel project!?

**Comment by szarnyasg** [comment]
Hi,

happy new year!

> Btw, currently the opencypher-xtext implementation is part of my slizaa project, but maybe it would make sense to promote it bit more and provide it in its own, stand-alone toplevel project!?

I think it'd definitely make sense to promote this as a standalone project. While [openCypher is adopted by more and more systems](https://www.slideshare.net/szarnyasg/compiling-opencypher-graph-queries-with-spark-catalyst/8), a full-fledged parser would be very useful for the graph community. (This is especially true as Xtext is no longer tied to the Eclipse ecosystem.)

This is also emphasised by the fact that the [Cypher for Apache Spark](https://github.com/opencypher/cypher-for-apache-spark/) project, developed under the openCypher umbrella, [uses Neo4j's Parboiled-based parser](https://github.com/opencypher/cypher-for-apache-spark/blob/0894660dbe7ef9dbbf1607f3c39bbf129882ecd6/caps-core/src/main/scala/org/opencypher/caps/impl/parse/CypherParser.scala#L18). I have not used [Parboiled](https://github.com/sirthias/parboiled) yet, but based on my understanding, it parses the code to an AST and developers can use rewriters to transform the tree. Compared to Xtext, Parboiled is less opinionated and offers more flexibility, but misses some of the more sophisticated features (e.g. generating an IDE with refactoring, etc.)

**Comment by szarnyasg** [comment]
I made some progress, but there are some things still to be fixed:

* [x] handle `WITH p AS p` and `RETURN p AS p` correctly
* [x] reintroduce keywords as they do not work correctly now (see also #11, which is loosely related)

**Comment by szarnyasg** [comment]
I fixed the two issues mentioned yesterday. I smoke tested the grammar manually (using the generated editor) and it seems fine, but more testing needs to be done.

**Comment by wuetherich** [comment]
HI Gabor,

I just pushed some unit tests to https://github.com/slizaa/slizaa-opencypher-xtext/tree/szarnyasg-grammar-scoping (https://github.com/slizaa/slizaa-opencypher-xtext/tree/szarnyasg-grammar-scoping/tests/org.slizaa.neo4j.opencypher.tests/src/org/slizaa/neo4j/opencypher/tests/scoping). The branch also contains your pull request.  

The first unit test parses a combined statement (* union *) and is ok. The second one parses and tests the statement you provided with this ticket and fails currently (it seems that the current implementation does not handle the 'overloading' or 'shading' correctly). However, it should not be such a big thing to fix this... I will take a closer look at your implementation tomorrow. 

Thanks again for your help and support!

**Comment by wuetherich** [comment]
Interestingly there are also several regression failures, as you can see here: https://travis-ci.org/slizaa/slizaa-opencypher-xtext/builds/326152939

E.g. the following statement is not handled correctly (games and both occurences of x couldn't be resolved):
```
START month = node:months('name:*')
MATCH (month)-[:in_month]-(game)-[:scored_in]-(player)
WITH month, player, count(game) AS games
ORDER BY games DESC
WITH month, 
    [x IN COLLECT([player.name, games])[0..5] | x[0]] AS players,
    [x IN COLLECT([player.name, games])[0..5] | x[1]] AS goals
RETURN month.name
```
Do you know if an 'official' specification for scoping in cypher exists?

**Comment by szarnyasg** [comment]
Thanks! I'll get back to this tomorrow in more detail.

> Do you know if an 'official' specification for scoping in cypher exists?

I'll ask around, but I don't think there is an official specification for this. Neo4j's Cypher engine does a lot of rewriting on the query ASTs, which eventually makes scopes trivial. Of course, we could use those rules to infer the scoping, but it is quite difficult (and not the most elegant approach). I studied the rewriters a bit when investigating issue https://github.com/neo4j/neo4j/issues/10712 - but in any case, rewriting the AST works well for a query engine, but does not work for an editor.
 
> The first unit test parses a combined statement (* union *) and is ok. The second one parses and tests the statement you provided with this ticket and fails currently (it seems that the current implementation does not handle the 'overloading' or 'shading' correctly

Yes, but this was intentional. To handle shadowing correctly, we would need to use nested scopes. While this could be done (and of course Xtext is perfectly up to the task), I found it to provide no additional value. But then again, I am not an expert in parser technologies.

 > E.g. the following statement is not handled correctly (games and both occurences of x couldn't be resolved)

Fixing 'games' was rather trivial. `x` will be a bit more difficult - so far I have only considered clauses that can introduce new variables. However, they can be introduced by other language constructs, such as list comprehensions, `FOREACH` and some functions (`filter`, `reduce`, `extract`, `all`, etc.).
  
Some work-in-progress code is at: https://github.com/szarnyasg/slizaa-opencypher-xtext/commit/d7eeb55555d850eb1fd6fa234f94b9f5d94e17a0

**Comment by szarnyasg** [comment]
Some notes: I think I fixed expressions (test are green). There are two test failures:

(1) The new `ScopeTest` class:
```
// Expected: [x, c, p, cities, p, c], but was: [p, c, cities, x]	
```

Currently, this behaviour is intentional in the scope provider. I'll think on this a bit more and discuss it with my expert colleagues.

(2) `FOREACH` is still broken as it can use nested clauses which are not yet handled. See this example (from class `ForEach_Test`), where `SET` is a subclause of `FOREACH`. This should not be too difficult to fix.

```
        Foreach 
            VariableDeclaration (name: n)
            FunctionInvocation (distinct: false)
                FunctionName (name: nodes)
                VariableRef  ->VariableDeclaration (name: p)
            Set 
                SetItem 
                    VariableRef  ->VariableDeclaration (eProxyURI: __synthetic0.cypher#|1)
                        PropertyLookup (propertyKeyName: marked, propertyOperator: null)
                    BoolConstant (value: TRUE)
```

**Comment by szarnyasg** [comment]
So as I promised, I had a discussion with my Xtext expert colleagues. A side effect of that was PR #16 :-).

W.r.t. to point (1), there are essentially two ways to approach scoping:

- Include definitions that could possibly be used there (e.g. all elements in the outer blocks in a usual structured programming language) and let shadowing do its job.
- Only include definitions that can be accessed - the approach I implemented.

The key tradeoff is that second approach cannot support sophisticated IDE features, such as quick fixes - for such features, it is often desirable to have all possible declarations (even if a declaration might not be accessed according to the language semantics, in which case it will be shadowed by the inner scopes or filtered by the linker). An example of such a quick fix feature:

```
MATCH (n)--(m)
WITH m
RETURN n.x, m.y
```

A quick fix could offer the user to include `n` in the `WITH` clause. While this would be very interesting and useful, this is probably future work. For now, am inclined to stop at implementing the scope provider.

**Comment by wuetherich** [comment]
I‘ll will have a look at this tomorrow...

**Comment by szarnyasg** [comment]
`MERGE ... ON ...` still has issues.

**Comment by szarnyasg** [comment]
Fixed in 1a37e3e00a091d6efbae0973743ec1113a269fd6.
